### ISSUE TITLE
Improper creation of OkHttpClient in Websocket.java probably causing Out Of Memory exceptions 

### ISSUE THREAD

**Issue by shobhitpuri** [socketio/engine.io-client-java#81]
**Issue**:
I have been having issue with 100's of Out of Memory exceptions in my Android app which is using [socket.io-client-java](https://github.com/socketio/socket.io-client-java) library since couple of months and I've been trying a lot to figure out. The library uses engine.io library.

**Hypothesis:**
After doing some exploration, this is the hypothesis I have. Please feel free to suggest your thoughts or a fix. In [Websocket.java](https://github.com/socketio/engine.io-client-java/blob/master/src/main/java/io/socket/engineio/client/transports/WebSocket.java) file, each time `doOpen()` is being called, a new instance of `OkHttpClient` is being initialized as follows:
```
public void doOpen() {
     // ..... 
     OkHttpClient.Builder clientBuilder = new OkHttpClient.Builder()
                // turn off timeouts (github.com/socketio/engine.io-client-java/issues/32)
                .connectTimeout(0, TimeUnit.MILLISECONDS)
                .readTimeout(0, TimeUnit.MILLISECONDS)
                .writeTimeout(0, TimeUnit.MILLISECONDS);
    // ...
    final OkHttpClient client = clientBuilder.build();
    // ...
}
```

Based on a comment by [swankjesse](https://github.com/swankjesse) at Square, on an [issue opened on okhttp](https://github.com/square/okhttp/issues/2846) regarding the OOM exceptions, he suggested to share the instance of `OkHttpClient`. Otherwise each time when we are creating a new instance of `OkHttpClient`, it will hold its own connection pool and thread pool, which is what seems to be happening when calling `doOpen()` in `Websocket.java`. An app which closes connection and opens connection multiple times in its workflow is having lot of OOM exceptions (stacktrace same as on https://github.com/square/okhttp/issues/2846). When a mobile app goes in background, one would want to close connection to prevent using resources and then re-open connection when app is visible. 

From the [okhttp docs](https://square.github.io/okhttp/3.x/okhttp/okhttp3/OkHttpClient.html ):
> OkHttp performs best when you create a single OkHttpClient instance and reuse it for all of your HTTP calls. This is because each client holds its own connection pool and thread pools. Reusing connections and threads reduces latency and saves memory. Conversely, creating a client for each request wastes resources on idle pools.

Solution seems to be fixing it as mentioned in the docs. Feel free to comment if the hypothesis seems wrong. 

Thanks

References:
1. https://github.com/square/okhttp/issues/2846
2. https://square.github.io/okhttp/3.x/okhttp/okhttp3/OkHttpClient.html 
3. Socket.io Java client use this library and it is causing OOM's in them. https://github.com/socketio/socket.io-client-java/issues/315

**Comment by b95505017** [comment]
Maybe we should let develop inject their OkHttpClient so an application could use only one client instance.

**Comment by Unchosen** [comment]
I am having this issue, in the case of my Socket.IO server being offline and the java client creating threads that do not close, forever. So I have attempted to make a fix in my own fork, which seems to be working for for my application.
With my changes, I am simply passing shareOkHttpClient as true to the socket options.
The way I did it may not sit well with others, hence why I have not created a pull request, but feel free to copy/use the same changes for the fix. https://github.com/Unchosen/engine.io-client-java/commit/51d41d64e98581efb90a6d30ddefaa3063552914

**Comment by shobhitpuri** [comment]
@Unchosen That's great news! I see that you forked the original repo. Can you please make a pull request to the original repo as well?

**Comment by shobhitpuri** [comment]
@b95505017 Can you please look into the pull request https://github.com/socketio/engine.io-client-java/pull/82 by @Unchosen when you get time? Thanks.

**Comment by b95505017** [comment]
@shobhitpuri Ok I'll review it ASAP.

**Comment by nkzawa** [comment]
The fix was released as 0.9.0, thanks to @b95505017. Let us know when the problem still happens.

**Comment by KevinLiu707** [comment]
[OOM still happended] (https://github.com/socketio/socket.io-client-java/issues/786)  @nkzawa  @shobhitpuri   . The real reason is  the threadpool closed when the task is done in EventThread class . But the same code is still in 0.9.0 nor later version.
